### PR TITLE
Wdio-validation

### PR DESCRIPTION
**Pull request by DaniMartinez-Gonzalez**
This pull request includes the webdriverio validation using the systelab components wdio library interacts with the https://systelab.github.io/components/.
Please, feel free to add suggestions and comments.

Thanks in advance for your time!

### CODE REVIEW

**Review by jserra91 — APPROVED** [review]

**Review by atzurk — APPROVED** [review]
Good job Dani! I'm not sure if here we should add some reporting on the AssertionUtility, just to confirm we are able to check and report all or most expectations/assertions.

File: app/wdio-validation/page-objects/showCaseComponentsPage.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+  public getTextBoxById(): InputField {
+    return new InputField(this.byId('full-width-input'));
+  }
+
+  public getIconQuestionButton(): Button {
+    return new Button(this.byCSS('.icon-question'));
+  }
+
+  public getTextBoxDisabledById(): Button {
```
Comment (JordiCela) [inline app/wdio-validation/page-objects/showCaseComponentsPage.ts]: The name of the method, maybe should be referred to a 'Button'.

Comment (Franek75) [reply app/wdio-validation/page-objects/showCaseComponentsPage.ts]: @JordiCela good catch!
But as the code inside is actually locating 'disabled-input', it appears that the method's name would be convenient, and it would be rather the return type and the constructor that might need to be changed to InputType.

File: app/wdio-validation/page-objects/showCaseComponentsPage.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
+import { BasePage, InputField, Label, Button } from 'systelab-components-wdio-test';
+
+export class ShowCaseComponentsPage extends BasePage {
+  constructor() {
+    super('showcase-components');
+  }
+
+  public getTextBoxByClassName(): InputField {
```
Comment (Franek75) [inline app/wdio-validation/page-objects/showCaseComponentsPage.ts:8]: Would it be convenient to rename getTextBox(....)() methods as getInputField(...)(), in order to keep more consistence with the actual name of the component type they are getting?

File: app/wdio-validation/page-objects/showCaseComponentsPage.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+  public getTextBoxByCss(): InputField {
+    return new InputField(this.byCSS('.form-control'));
+  }
+
+  public getTextBoxById(): InputField {
+    return new InputField(this.byId('full-width-input'));
+  }
+
+  public getIconQuestionButton(): Button {
```
Comment (Franek75) [inline app/wdio-validation/page-objects/showCaseComponentsPage.ts:20]: getIconQuestionButtonByCSS()? (for consistence with the analogue methods for other components)

File: app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,148 @@
+import { AssertionUtility, ReportUtility, TestIdentification, DefaultTimeout } from "systelab-components-wdio-test";
+import { BrowserInteractionService } from "../services/browserInteraction.service";
+import { ShowCaseComponentsPage } from "../page-objects/showCaseComponentsPage";
+import { SystelabModalComponent } from "../page-objects/systelabModalComponent";
+
+describe("TC0001_WebdriverIO-Validation_Action_OnSelectors", () => {
+  let browserService: BrowserInteractionService;
+  let showCasePage: ShowCaseComponentsPage;
+  let modalPage: SystelabModalComponent;
+  let systelabTitle = "Systelab Components Library";
```
Comment (Franek75) [inline app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts]: Const instead of let? (also for the following declarations)

File: app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    TestIdentification.setDescription(
+      "Goal: The purpose of this test case is to verify different actions performed on" +
+        "elements of the Systelab Components website on different browsers"
+    );
+    await browserService.navigateToSystelabComponents();
+  });
+
+  it("TC0001-01-WebdriverIO-Validation - Navigate to Systelab Components, find an Input Field, set and get a value", async () => {
+    await showCasePage.waitToBeDisplayed();
+    await ReportUtility.addExpectedResult(
```
Comment (Franek75) [inline app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts]: As this code is used in every test of the suite, it is a good candidate to be encapsulated inside an expectation service method (e.g. expectWebsiteOpenAndActive)
Just as a suggestion.

File: app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+      async () => {
+        AssertionUtility.expectContains(
+          await browser.getTitle(),
+          systelabTitle
+        );
+      }
+    );
+
+    await showCasePage.getTextBoxById().setText(foundByID);
+    await ReportUtility.addExpectedResult(
```
Comment (Franek75) [inline app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts:54]: Also this one could be encapsulated as an expectation service method (it is present in 2 tests), e.g. expectInputFieldValueCorrectlySet()
Just as a suggestion

File: app/wdio-validation/page-objects/showCaseComponentsPage.ts (hunk 1 above)
Comment (DaniMartinez-Gonzalez) [reply app/wdio-validation/page-objects/showCaseComponentsPage.ts]: Reviewing that part of the code we figure out that this method was duplicated and unused. 
Now it would be fixed.

Thanks a lot! :-)

File: app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,148 @@
+import { AssertionUtility, ReportUtility, TestIdentification, DefaultTimeout } from "systelab-components-wdio-test";
+import { BrowserInteractionService } from "../services/browserInteraction.service";
+import { ShowCaseComponentsPage } from "../page-objects/showCaseComponentsPage";
+import { SystelabModalComponent } from "../page-objects/systelabModalComponent";
+
+describe("TC0001_WebdriverIO-Validation_Action_OnSelectors", () => {
+  let browserService: BrowserInteractionService;
+  let showCasePage: ShowCaseComponentsPage;
+  let modalPage: SystelabModalComponent;
```
Comment (Franek75) [inline app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts]: The name might be a little confusing, I mean when I was reading the code below I thought it was a page (a PO) and then I recalled that it is a Widget.
I would rename it just to leave the ...Page "suffix" exclusive for instances of POs. Just a suggestion.

File: app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+      "The Systelab Components website is open and active",
+      async () => {
+        AssertionUtility.expectContains(
+          await browser.getTitle(),
+          systelabTitle
+        );
+      }
+    );
+    await showCasePage.getIconQuestionButton().click();
+    await modalPage.waitToBeDisplayed(DefaultTimeout.SLOW_WAIT);
```
Comment (Franek75) [inline app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts]: If SLOW_WAIT is going to be the default timeout, it could be set rather inside waitToBeDisplayed() (overriding the method inherited from Widget) and this way to get rid from passing the parameter, at least in the default cases.
This is just a suggestion, as I am not that familiar with the global variety of uses this is intended to.

File: app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        AssertionUtility.expectContains(
+          await browser.getTitle(),
+          systelabTitle
+        );
+      }
+    );
+    await showCasePage.getTextBoxDisabled().setText(disabledTextBox);
+    await showCasePage.getTextBoxById().setText(enabledTextBox);
+    await ReportUtility.addExpectedResult(
+      "On disabled input field is not been able to set value",
```
Comment (Franek75) [inline app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts]: I would reword this, because you are actually adding 2 expectations and somehow comparing their results (values can be set on enabled input fields, values can not be set on disabled input fields), thus the reported result might contain something about the enabled input field also

File: app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+    await modalPage.waitToBeDisplayed(DefaultTimeout.SLOW_WAIT);
+    await ReportUtility.addExpectedResult(
+      "When perform a click on a button of the systelab component website, an expected modal window is open",
+      async () => {
+        AssertionUtility.expectTrue(await modalPage.isDisplayed());
+      }
+    );
+  });
+
+  it("TC0001-04-WebdriverIO-Validation - Navigate to Systelab Components, on disabled Input Field try to set a value", async () => {
```
Comment (Franek75) [inline app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts]: Implementation performs actions on both the enabled and the disabled Input Field, thus it would be nice if the it description told about the "enabled Input Field" also.
Or, if it is preferred to focus on the disabled Input Field behavior, the "enabled Input Field"-related actions+expectations could be dropped)

**Review by Franek75 — APPROVED** [review]
Looks like a very good job! I did not find nothing wrong. Please take all my comments as suggestions

File: app/wdio-validation/page-objects/showCaseComponentsPage.ts (hunk 2 above)
Comment (DaniMartinez-Gonzalez) [reply app/wdio-validation/page-objects/showCaseComponentsPage.ts:8]: I understand you're comment, and I'm totally agreed, but for validation review purposes we thought that could be more readable and better for understanding it.
That's why we don't put these methods to getInputFieldBy..

We could talk if you want.

Thanks Fran for your comments :-)

image

File: app/wdio-validation/page-objects/showCaseComponentsPage.ts (hunk 3 above)
Comment (DaniMartinez-Gonzalez) [reply app/wdio-validation/page-objects/showCaseComponentsPage.ts:20]: The rest of the methods that were created with a byId, byClassName, by TagName are focused to verify that the same element (our input field or textbox) can be found in different ways. 

The iconQuestion button is used for another type of test.

File: app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts (hunk 4 above)
Comment (DaniMartinez-Gonzalez) [reply app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts]: You're right! thanks

File: app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts (hunk 8 above)
Comment (DaniMartinez-Gonzalez) [reply app/wdio-validation/specs/actionInteraction-wdio.e2e-spec.ts]: In that case, we have to pass the SLOW_WAIT (5000ms) because when the test clicks on a button then the 500ms default time (FAST_WAIT) it wasn't enough. The modal is not displayed yet.

File: app/wdio-validation/page-objects/showCaseComponentsPage.ts (hunk 1 above)
Comment (JordiCela) [reply app/wdio-validation/page-objects/showCaseComponentsPage.ts]: You are welcome ;)

**Review by JordiCela — APPROVED** [review]

**Review by jserra91 — APPROVED** [review]

**Review by oshura — APPROVED** [review]
